File: scripts/sync-from-numbers.py
#!/usr/bin/env python3
"""
sync-from-numbers.py

Reads "2025-26 renewals" and "property info" sheets directly from the
Numbers file on Google Drive and upserts everything into Supabase.

Sensitive columns are never read:
  - col 21 (Door Codes)
  - col 26 (Lock Box and Key Number)

Usage:
  python3 scripts/sync-from-numbers.py
"""

import os, re, sys, json, datetime, requests
from pathlib import Path
from numbers_parser import Document

# ─── Config ───────────────────────────────────────────────────────────────────

NUMBERS_FILE = "/Volumes/One Touch/The_Team_Google_Drive Sync/2025-2026 Renewals_Dashboard.numbers"

SHEET_RENEWALS  = "2025-26 renewals"
SHEET_PROP_INFO = "property info"

# Load .env from project root
env_path = Path(__file__).parent.parent / ".env"
env = {}
for line in env_path.read_text().splitlines():
    line = line.strip()
    if line and not line.startswith("#") and "=" in line:
        k, _, v = line.partition("=")
        env[k.strip()] = v.strip()

SUPABASE_URL = env.get("SUPABASE_URL", "")
SUPABASE_KEY = env.get("SUPABASE_SERVICE_KEY", "")

if not SUPABASE_URL or not SUPABASE_KEY:
    print("ERROR: Missing SUPABASE_URL or SUPABASE_SERVICE_KEY in .env")
    sys.exit(1)

HEADERS = {
    "apikey": SUPABASE_KEY,
    "Authorization": f"Bearer {SUPABASE_KEY}",
    "Content-Type": "application/json",
    "Prefer": "resolution=merge-duplicates,return=minimal",
}

# ─── Supabase helpers ──────────────────────────────────────────────────────────

def sb_upsert(table, rows, on_conflict="address"):
    if not rows:
        return
    url = f"{SUPABASE_URL}/rest/v1/{table}"
    params = {"on_conflict": on_conflict}
    r = requests.post(url, headers=HEADERS, params=params, json=rows)
    if r.status_code not in (200, 201):
        print(f"  ERROR upserting {table}: {r.status_code} {r.text[:200]}")
        sys.exit(1)

def sb_delete_in(table, column, ids):
    if not ids:
        return
    url = f"{SUPABASE_URL}/rest/v1/{table}"
    id_list = ",".join(f'"{i}"' if isinstance(i, str) else str(i) for i in ids)
    r = requests.delete(url, headers=HEADERS, params={column: f"in.({id_list})"})
    if r.status_code not in (200, 204):
        print(f"  ERROR deleting from {table}: {r.status_code} {r.text[:200]}")

def sb_select(table, columns="*"):
    url = f"{SUPABASE_URL}/rest/v1/{table}"
    r = requests.get(url, headers={**HEADERS, "Prefer": "return=representation"},
                     params={"select": columns})
    if r.status_code != 200:
        print(f"  ERROR fetching {table}: {r.status_code} {r.text[:200]}")
        sys.exit(1)
    return r.json()

def sb_insert(table, rows):
    if not rows:
        return
    url = f"{SUPABASE_URL}/rest/v1/{table}"
    r = requests.post(url, headers={**HEADERS, "Prefer": "resolution=ignore-duplicates,return=minimal"},
                      json=rows)
    if r.status_code not in (200, 201):
        print(f"  ERROR inserting {table}: {r.status_code} {r.text[:200]}")
        sys.exit(1)

# ─── Data helpers ──────────────────────────────────────────────────────────────

def cell_str(cell):
    """Return cell value as a clean string."""
    v = cell.value
    if v is None:
        return ""
    if isinstance(v, datetime.datetime):
        return v.date().isoformat()        # → "2026-07-31"
    if isinstance(v, datetime.date):
        return v.isoformat()
    s = str(v).strip()
    return "" if re.match(r'^[—–\-]+$', s) else s

def yn(cell):
    return cell_str(cell).lower() == "yes"

def parse_date(cell):
    s = cell_str(cell)
    if not s:
        return None
    # Already ISO from Numbers datetime: "2026-07-31"
    if re.match(r'^\d{4}-\d{2}-\d{2}$', s):
        return s
    # Fallback MM/DD/YY or MM/DD/YYYY
    parts = s.split("/")
    if len(parts) == 3:
        m, d, y = (int(p) for p in parts)
        year = 2000 + y if y < 100 else y
        return f"{year}-{m:02d}-{d:02d}"
    return None

def parse_beds(cell):
    s = cell_str(cell)
    if not s:
        return None
    # Numbers stores numeric values as floats: "2.0" → "2"
    try:
        f = float(s)
        return str(int(f)) if f == int(f) else s
    except ValueError:
        return s  # e.g. "Studio", "1, with an office"

# ─── Load Numbers file ─────────────────────────────────────────────────────────

if not Path(NUMBERS_FILE).exists():
    print(f"ERROR: Numbers file not found at:\n  {NUMBERS_FILE}")
    print("Is the external drive / Google Drive volume mounted?")
    sys.exit(1)

print(f"Reading: {Path(NUMBERS_FILE).name}")
doc = Document(NUMBERS_FILE)

sheets_by_name = {s.name: s for s in doc.sheets}
if SHEET_RENEWALS not in sheets_by_name:
    print(f"ERROR: Sheet '{SHEET_RENEWALS}' not found. Available: {list(sheets_by_name)}")
    sys.exit(1)
if SHEET_PROP_INFO not in sheets_by_name:
    print(f"ERROR: Sheet '{SHEET_PROP_INFO}' not found. Available: {list(sheets_by_name)}")
    sys.exit(1)

# ─── Parse "property info" ─────────────────────────────────────────────────────
# Col indices (0-based):
#  0=Property  1=Beds  2=Baths  6=Town  7=Type  8=SqFt  10=Utilities
#  11=Freeze   18=Pets  22=Owner  23=Portfolio  24=Area  25=Notes
#  21=DoorCodes (SKIP)  26=LockBox (SKIP)

prop_rows = list(sheets_by_name[SHEET_PROP_INFO].tables[0].iter_rows())
prop_info = {}  # address → dict

for row in prop_rows[1:]:   # skip header
    addr = cell_str(row[0])
    if not addr:
        continue
    prop_info[addr] = {
        "address":       addr,
        "beds":          parse_beds(row[1]),
        "baths":         float(cell_str(row[2])) if cell_str(row[2]) else None,
        "property_type": cell_str(row[7]),
        "sq_ft":         int(float(cell_str(row[8]))) if cell_str(row[8]) else None,
        "utilities":     cell_str(row[10]),
        "freeze_warning": yn(row[11]),
        "pets_allowed":  cell_str(row[18]),
        "owner_name":    cell_str(row[22]),
        "area":          cell_str(row[24]),
        # cols 21 (door codes) and 26 (lock box) deliberately not read
    }

print(f"  property info: {len(prop_info)} properties")

# ─── Parse "2025-26 renewals" ──────────────────────────────────────────────────
# Col indices (0-based):
#  0=Property  1=Resident  2=Email  3=LeaseEnd  4=Status
#  5=LeaseSigned  6=DepositPaid  7=Notes
#  8=NextResident  9=NextEmail  10=NextPhone  11=NextLeaseEnd
#  14=Owner  15=Area

renewal_rows = list(sheets_by_name[SHEET_RENEWALS].tables[0].iter_rows())
groups = {}  # address → list of raw rows

for row in renewal_rows[1:]:   # skip header
    addr = cell_str(row[0])
    if not addr:
        continue
    resident = cell_str(row[1])
    if resident.lower() == "airbnb":
        continue
    if addr not in groups:
        groups[addr] = []
    groups[addr].append(row)

print(f"  renewals: {len(groups)} properties, "
      f"{sum(len(v) for v in groups.values())} resident rows")

# ─── Build unit upsert rows ───────────────────────────────────────────────────

unit_upserts = []
for addr, rows in groups.items():
    info = prop_info.get(addr, {})
    owner = next((cell_str(r[14]) for r in rows if cell_str(r[14])), "") or info.get("owner_name", "")
    area  = next((cell_str(r[15]) for r in rows if cell_str(r[15])), "") or info.get("area", "")
    unit_upserts.append({
        "address":       addr,
        "beds":          info.get("beds"),
        "baths":         info.get("baths"),
        "area":          area,
        "owner_name":    owner,
        "utilities":     info.get("utilities", ""),
        "property_type": info.get("property_type", ""),
        "sq_ft":         info.get("sq_ft"),
        "freeze_warning": info.get("freeze_warning", False),
        "pets_allowed":  info.get("pets_allowed", ""),
    })

# Properties in property info with no current residents
for addr, info in prop_info.items():
    if addr not in groups:
        unit_upserts.append({
            "address":       addr,
            "beds":          info.get("beds"),
            "baths":         info.get("baths"),
            "area":          info.get("area", ""),
            "owner_name":    info.get("owner_name", ""),
            "utilities":     info.get("utilities", ""),
            "property_type": info.get("property_type", ""),
            "sq_ft":         info.get("sq_ft"),
            "freeze_warning": info.get("freeze_warning", False),
            "pets_allowed":  info.get("pets_allowed", ""),
        })

# ─── Upsert units ─────────────────────────────────────────────────────────────
print(f"\nUpserting {len(unit_upserts)} units...")
sb_upsert("units", unit_upserts, on_conflict="address")
print(f"  ✓ {len(unit_upserts)} units")

# ─── Fetch unit ID map ────────────────────────────────────────────────────────
unit_data = sb_select("units", "id,address")
unit_id_map = {u["address"]: u["id"] for u in unit_data}

# ─── Clear + reinsert residents for synced addresses ─────────────────────────
synced_ids = [unit_id_map[a] for a in groups if a in unit_id_map]
print(f"Clearing residents for {len(synced_ids)} units...")
sb_delete_in("residents",      "unit_id", synced_ids)
sb_delete_in("next_residents", "unit_id", synced_ids)

# ─── Build resident rows ──────────────────────────────────────────────────────
resident_rows      = []
next_resident_rows = []

for addr, rows in groups.items():
    unit_id = unit_id_map.get(addr)
    if not unit_id:
        print(f"  ⚠ No unit ID for: {addr}")
        continue

    seen_next = set()
    for row in rows:
        name = cell_str(row[1])
        if not name:
            continue

        resident_rows.append({
            "unit_id":     unit_id,
            "name":        name,
            "email":       cell_str(row[2]),
            "lease_end":   parse_date(row[3]),
            "status":      cell_str(row[4]).lower() or "unknown",
            "lease_signed": yn(row[5]),
            "deposit_paid": yn(row[6]),
            "notes":       cell_str(row[7]),
        })

        next_name  = cell_str(row[8])
        next_email = cell_str(row[9])
        if next_name:
            key = next_email or next_name
            if key not in seen_next:
                seen_next.add(key)
                next_resident_rows.append({
                    "unit_id": unit_id,
                    "name":    next_name,
                    "email":   next_email,
                    "phone":   cell_str(row[10]),
                })

print(f"Inserting {len(resident_rows)} residents...")
sb_insert("residents", resident_rows)
print(f"  ✓ {len(resident_rows)} residents")

if next_resident_rows:
    print(f"Inserting {len(next_resident_rows)} next residents...")
    sb_insert("next_residents", next_resident_rows)
    print(f"  ✓ {len(next_resident_rows)} next residents")

# ─── Log the sync ─────────────────────────────────────────────────────────────
sb_insert("sync_log", [{
    "source":              "numbers",
    "status":              "success",
    "units_upserted":      len(unit_upserts),
    "residents_upserted":  len(resident_rows),
}])

print("\n✅ Sync complete.")

File: scripts/sync-from-numbers.mjs
/**
 * sync-from-numbers.mjs
 * Reads exported CSVs from the Numbers file and upserts into Supabase.
 * Run: node scripts/sync-from-numbers.mjs
 *
 * Expects env vars: SUPABASE_URL, SUPABASE_SERVICE_KEY
 * CSV paths passed as args or defaults below.
 */

import { createClient } from '@supabase/supabase-js';
import { readFileSync, mkdirSync } from 'fs';
import { resolve } from 'path';
import { execSync } from 'child_process';

const SUPABASE_URL = process.env.SUPABASE_URL;
const SUPABASE_SERVICE_KEY = process.env.SUPABASE_SERVICE_KEY;

if (!SUPABASE_URL || !SUPABASE_SERVICE_KEY) {
  console.error('Error: SUPABASE_URL and SUPABASE_SERVICE_KEY must be set in .env');
  process.exit(1);
}

// ─── Column indices (updated April 2026 — Move Out moved to col 8, new col 13 inserted) ─
const S1 = {
  PROPERTY:     0,
  RESIDENT:     1,
  EMAIL:        2,
  PHONE:        3,
  LEASE_END:    4,
  STATUS:       5,  // was 6
  LEASE_SIGNED: 6,  // was 7
  DEPOSIT_PAID: 7,  // was 8
  MOVE_OUT:     8,  // was 5 (col moved; "Next year's lease end date" inserted at 13)
  NOTES:        9,
  NEXT_RESIDENT:10,
  NEXT_EMAIL:   11,
  NEXT_PHONE:   12,
  NEXT_MOVE_IN: 14, // was 13 (shifted by new col 13 "Next year's lease end date")
  OWNER:        17,
  AREA:         18,
};

// S2 (Sheet 2 / property info) removed — property attributes now owned by
// the Property Info Google Sheet and synced to Supabase via sync-property-cache.mjs.

// ─── Helpers ─────────────────────────────────────────────────────────────────

// Normalize addresses for matching: lowercase, strip periods, collapse spaces, trim.
// "201 E. Carr St" and "201 E Carr St" → "201 e carr st"
function normalizeAddr(addr) {
  return (addr || '').toLowerCase().replace(/\./g, '').replace(/\s+/g, ' ').trim();
}

// Strip common street suffixes for fuzzy matching.
// "230 Valley Park Dr" → "230 valley park", "110 Fidelity St" → "110 fidelity"
const STREET_SUFFIXES = /\s+(dr|st|rd|ave|blvd|ct|ln|cir|way|pl)\.?$/i;
function stripSuffix(addr) {
  return normalizeAddr(addr).replace(STREET_SUFFIXES, '');
}

function parseCSV(text) {
  const rows = [];
  for (const line of text.split('\n')) {
    if (!line.trim()) continue;
    const row = [];
    let inQuote = false, cell = '';
    for (let i = 0; i < line.length; i++) {
      const ch = line[i];
      if (ch === '"') { inQuote = !inQuote; }
      else if (ch === ',' && !inQuote) { row.push(cell.trim()); cell = ''; }
      else { cell += ch; }
    }
    row.push(cell.trim());
    rows.push(row);
  }
  return rows;
}

function clean(val) {
  if (!val) return '';
  const s = val.toString().trim();
  return /^[—–-]+$/.test(s) ? '' : s;
}

function yn(val) {
  return (val || '').toString().trim().toLowerCase() === 'yes';
}

function parseDate(val) {
  const s = clean(val);
  if (!s) return null;
  // ISO from numbers-parser: "2026-07-31 00:00:00" or "2026-07-31"
  const isoMatch = s.match(/^(\d{4})-(\d{2})-(\d{2})/);
  if (isoMatch) return `${isoMatch[1]}-${isoMatch[2]}-${isoMatch[3]}`;
  // M/D/YY from manual CSV export
  const [m, d, y] = s.split('/').map(Number);
  if (!m || !d || !y) return null;
  const year = y < 100 ? 2000 + y : y;
  return `${year}-${String(m).padStart(2,'0')}-${String(d).padStart(2,'0')}`;
}

// ─── Export from Numbers via numbers-parser (Python) ─────────────────────────
// NUMBERS_FILE env var allows cloud runners (GitHub Actions) to pass a downloaded path.
// Falls back to the local Google Drive mount for Mac Mini scheduled runs.
const NUMBERS_FILE = process.env.NUMBERS_FILE
  || '/Volumes/One Touch/The_Team_Google_Drive Sync/2025-2026 Renewals_Dashboard.numbers';
const EXPORT_DIR = '/tmp/mills_export';

function exportFromNumbers() {
  console.log('Reading Numbers file directly...');
  mkdirSync(EXPORT_DIR, { recursive: true });

  const pyScript = `
import numbers_parser, csv, os

doc = numbers_parser.Document("${NUMBERS_FILE}")
sheet = doc.sheets[0]  # Sheet 1 only — renewals/tenant data
rows = sheet.tables[0].rows(values_only=True)
rows = [['' if v is None else str(v) for v in row] for row in rows]
fname = os.path.join("${EXPORT_DIR}", "2025-26 renewals.csv")
with open(fname, "w", newline="", encoding="utf-8") as f:
    csv.writer(f).writerows(rows)
print(f"  Wrote {len(rows)} rows -> {fname}")
`;

  try {
    const out = execSync(`python3 -c '${pyScript.replace(/'/g, "'\\''")}'`, {
      timeout: 30000,
      encoding: 'utf8',
    });
    process.stdout.write(out);
    console.log('  ✓ Export complete');
  } catch (err) {
    console.error('  ✗ numbers-parser export failed:', err.stderr || err.message);
    console.error('  Falling back to existing CSVs in', EXPORT_DIR);
  }
}

exportFromNumbers();

// ─── Main ────────────────────────────────────────────────────────────────────
const sheet1Path = process.argv[2] || `${EXPORT_DIR}/2025-26 renewals.csv`;

console.log('Reading CSV...');
const sheet1Csv = readFileSync(resolve(sheet1Path), 'utf8');
const sheet1Rows = parseCSV(sheet1Csv);
console.log(`  Sheet1: ${sheet1Rows.length - 1} data rows`);

const sb = createClient(SUPABASE_URL, SUPABASE_SERVICE_KEY);

// Group Sheet1 rows by address
const groups = new Map();
for (let i = 1; i < sheet1Rows.length; i++) {
  const row = sheet1Rows[i];
  const addr = clean(row[S1.PROPERTY]);
  if (!addr) continue;
  if ((row[S1.RESIDENT] || '').trim().toLowerCase() === 'airbnb') continue;
  if (!groups.has(addr)) groups.set(addr, []);
  groups.get(addr).push(row);
}

console.log(`\nFound ${groups.size} units in Sheet1`);

// Fetch existing units from Supabase — property info sheet is the authoritative
// source for what units exist. Numbers sync never creates new unit rows.
const { data: unitData, error: fetchErr } = await sb.from('units').select('id, address, owner_name, area');
if (fetchErr) { console.error('fetch units failed:', fetchErr.message); process.exit(1); }

// Build three lookup maps: exact → id, normalized → id, suffix-stripped → id
const unitIdMap = Object.fromEntries(unitData.map(u => [u.address, u.id]));
const normalizedIdMap = Object.fromEntries(unitData.map(u => [normalizeAddr(u.address), u.id]));
const strippedIdMap = Object.fromEntries(unitData.map(u => [stripSuffix(u.address), u.id]));

// Three-tier address resolver: exact → normalized → suffix-stripped
function resolveUnitId(address) {
  return unitIdMap[address]
    ?? normalizedIdMap[normalizeAddr(address)]
    ?? strippedIdMap[stripSuffix(address)];
}

// Update owner_name and area for matched units (these come from the Numbers file)
const ownerUpdates = [];
for (const [address, rows] of groups) {
  const owner = rows.map(r => clean(r[S1.OWNER])).find(v => v) || '';
  const area  = rows.map(r => clean(r[S1.AREA])).find(v => v) || '';
  const id = resolveUnitId(address);
  if (id && (owner || area)) ownerUpdates.push({ id, owner_name: owner || undefined, area: area || undefined });
}
if (ownerUpdates.length > 0) {
  for (const { id, ...fields } of ownerUpdates) {
    await sb.from('units').update(fields).eq('id', id);
  }
  console.log(`  ✓ Updated owner/area for ${ownerUpdates.length} units`);
}

// Resolve Numbers addresses to existing Supabase unit IDs (three-tier matching)
const unmatched = [];
const resolvedIds = new Map(); // Numbers address → Supabase unit id
for (const address of groups.keys()) {
  const id = resolveUnitId(address);
  if (id) {
    resolvedIds.set(address, id);
  } else {
    unmatched.push(address);
  }
}
if (unmatched.length > 0) {
  console.warn(`\n⚠️  ${unmatched.length} address(es) from Numbers not found in Supabase (skipping residents):`);
  unmatched.forEach(a => console.warn(`   "${a}"`));
  console.warn('  Fix: update the address in Amanda\'s Numbers file to match the Property Info sheet exactly.\n');
}

// Snapshot residents + next_residents before deleting — used for change-detection email
const syncedUnitIds = [...resolvedIds.values()];
const [{ data: snapshotResidents }, { data: snapshotNextResidents }] = syncedUnitIds.length > 0
  ? await Promise.all([
      sb.from('residents').select('unit_id, name, status, lease_signed, deposit_paid').in('unit_id', syncedUnitIds),
      sb.from('next_residents').select('unit_id, name, email, phone, move_in_date').in('unit_id', syncedUnitIds),
    ])
  : [{ data: [] }, { data: [] }];

// Delete existing residents/next_residents for synced units
if (syncedUnitIds.length > 0) {
  await sb.from('residents').delete().in('unit_id', syncedUnitIds);
  await sb.from('next_residents').delete().in('unit_id', syncedUnitIds);
}

// Build resident rows
const residentRows = [];
const nextResidentRows = [];

for (const [address, rows] of groups) {
  const unitId = resolvedIds.get(address);
  if (!unitId) continue;
  const seenKeys = new Set();

  for (const row of rows) {
    const name = clean(row[S1.RESIDENT]);
    if (!name) continue;
    residentRows.push({
      unit_id: unitId,
      name,
      email: clean(row[S1.EMAIL]),
      phone: clean(row[S1.PHONE]),
      status: clean(row[S1.STATUS]).toLowerCase() || 'unknown',
      lease_end: parseDate(row[S1.LEASE_END]),
      move_out_date: parseDate(row[S1.MOVE_OUT]),
      lease_signed: yn(row[S1.LEASE_SIGNED]),
      deposit_paid: yn(row[S1.DEPOSIT_PAID]),
      notes: clean(row[S1.NOTES]),
    });

    const nextName = clean(row[S1.NEXT_RESIDENT]);
    const nextEmail = clean(row[S1.NEXT_EMAIL]);
    if (nextName) {
      const key = nextEmail || nextName;
      if (!seenKeys.has(key)) {
        seenKeys.add(key);
        nextResidentRows.push({
          unit_id: unitId,
          name: nextName,
          email: nextEmail,
          phone: clean(row[S1.NEXT_PHONE]),
          move_in_date: parseDate(row[S1.NEXT_MOVE_IN]),
        });
      }
    }
  }
}

console.log(`Inserting ${residentRows.length} residents...`);
if (residentRows.length > 0) {
  const { error: resErr } = await sb.from('residents').insert(residentRows);
  if (resErr) {
    if (resErr.message.includes('column') && resErr.message.includes('does not exist')) {
      console.error('\n⚠️  Missing columns in Supabase. Run this SQL in your Supabase SQL editor:');
      console.error(`
ALTER TABLE residents ADD COLUMN IF NOT EXISTS phone text DEFAULT '';
ALTER TABLE residents ADD COLUMN IF NOT EXISTS move_out_date date;
ALTER TABLE next_residents ADD COLUMN IF NOT EXISTS move_in_date date;
`);
      console.error('Then re-run this script.\n');
    } else {
      console.error('residents insert failed:', resErr.message);
    }
    process.exit(1);
  }
  console.log('  ✓ residents done');
}

console.log(`Inserting ${nextResidentRows.length} next_residents...`);
if (nextResidentRows.length > 0) {
  const { error: nrErr } = await sb.from('next_residents').insert(nextResidentRows);
  if (nrErr) {
    if (nrErr.message.includes('column') && nrErr.message.includes('does not exist')) {
      console.error('\n⚠️  Missing columns in Supabase. Run this SQL in your Supabase SQL editor:');
      console.error(`
ALTER TABLE residents ADD COLUMN IF NOT EXISTS phone text DEFAULT '';
ALTER TABLE residents ADD COLUMN IF NOT EXISTS move_out_date date;
ALTER TABLE next_residents ADD COLUMN IF NOT EXISTS move_in_date date;
`);
      console.error('Then re-run this script.\n');
    } else {
      console.error('next_residents insert failed:', nrErr.message);
    }
    process.exit(1);
  }
  console.log('  ✓ next_residents done');
}

console.log(`\n✓ Sync complete — ${resolvedIds.size} units matched, ${residentRows.length} residents, ${nextResidentRows.length} next residents`);

// ─── Change detection + email ─────────────────────────────────────────────────
// Only runs if GMAIL_USER + GMAIL_APP_PASSWORD + MEETING_EMAIL_TO are set.
// Safe to skip — missing env vars = no email, no error.

const GMAIL_USER = process.env.GMAIL_USER;
const GMAIL_PASS = process.env.GMAIL_APP_PASSWORD;
const EMAIL_TO   = process.env.MEETING_EMAIL_TO;

if (GMAIL_USER && GMAIL_PASS && EMAIL_TO) {
  // Fetch freshly-inserted data for diffing
  const [{ data: newResidents }, { data: newNextResidents }] = await Promise.all([
    sb.from('residents').select('unit_id, name, status, lease_signed, deposit_paid').in('unit_id', syncedUnitIds),
    sb.from('next_residents').select('unit_id, name, email, phone, move_in_date').in('unit_id', syncedUnitIds),
  ]);

  const changes = [];

  // --- Residents diff ---
  const prevMap  = Object.fromEntries((snapshotResidents || []).map(r => [`${r.unit_id}:${r.name}`, r]));
  const afterMap = Object.fromEntries((newResidents || []).map(r => [`${r.unit_id}:${r.name}`, r]));

  for (const [key, r] of Object.entries(afterMap)) {
    if (!prevMap[key]) changes.push(`+ Added resident: ${r.name}`);
  }
  for (const [key, r] of Object.entries(prevMap)) {
    if (!afterMap[key]) changes.push(`- Removed resident: ${r.name}`);
  }
  for (const [key, after] of Object.entries(afterMap)) {
    const before = prevMap[key];
    if (!before) continue;
    if (before.status !== after.status)
      changes.push(`~ ${after.name}: status ${before.status} → ${after.status}`);
    if (before.lease_signed !== after.lease_signed)
      changes.push(`~ ${after.name}: lease_signed → ${after.lease_signed}`);
    if (before.deposit_paid !== after.deposit_paid)
      changes.push(`~ ${after.name}: deposit_paid → ${after.deposit_paid}`);
  }

  // --- Next residents diff ---
  const prevNext  = Object.fromEntries((snapshotNextResidents || []).map(r => [`${r.unit_id}:${r.name}`, r]));
  const afterNext = Object.fromEntries((newNextResidents || []).map(r => [`${r.unit_id}:${r.name}`, r]));

  for (const [key, r] of Object.entries(afterNext)) {
    if (!prevNext[key]) changes.push(`+ Added next resident: ${r.name}`);
  }
  for (const [key, r] of Object.entries(prevNext)) {
    if (!afterNext[key]) changes.push(`- Removed next resident: ${r.name}`);
  }
  for (const [key, after] of Object.entries(afterNext)) {
    const before = prevNext[key];
    if (!before) continue;
    if (before.move_in_date !== after.move_in_date)
      changes.push(`~ ${after.name}: move_in_date ${before.move_in_date} → ${after.move_in_date}`);
  }

  if (changes.length > 0) {
    const { createTransport } = await import('nodemailer');
    const transporter = createTransport({
      service: 'gmail',
      auth: { user: GMAIL_USER, pass: GMAIL_PASS },
    });
    await transporter.sendMail({
      from: GMAIL_USER,
      to: EMAIL_TO,
      subject: `Mills Sync — ${changes.length} change${changes.length !== 1 ? 's' : ''} (${new Date().toLocaleDateString('en-US', { month: 'short', day: 'numeric' })})`,
      text: [
        `Numbers sync ran at ${new Date().toLocaleString('en-US', { timeZone: 'America/New_York' })} ET`,
        '',
        ...changes,
        '',
        `${resolvedIds.size} units matched · ${residentRows.length} residents · ${nextResidentRows.length} next residents`,
      ].join('\n'),
    });
    console.log(`  ✉ Change email sent (${changes.length} changes)`);
  } else {
    console.log('  ✉ No changes — skipping email');
  }
}
